---
title: "motifs"
author: "Tim Stuart"
date: "18/10/2016"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Motifs


Aim: to determine the proportion of TF motifs within each TE family, and the proportion of TE family members containing motifs.

I'm using the E2F TF, as it was previously shown to be enriched within TEs in Arabidopsis (Henaff et al. 2014 The Plant Journal). See previous document looking at peaks called within TEs.

```{r import, message=FALSE}
library(BSgenome.Athaliana.TAIR.TAIR9)
library(seqLogo)
library(Biostrings)
library(rtracklayer)
library(readr)
library(dplyr)
library(ggplot2)
```

### Position weight matrix

Need to create a position weight matrix from the MEME file for E2FA.

```{r}
meme <- as.matrix(read_tsv("../RawData/motifs/E2FDP_tnt/DEL1_col_a/meme_out/meme_m1.txt", skip = 12,
                 col_names = c("A", "C", "G", "T")))
pwm <- t(meme)
pwm
```

Draw a logo:

```{r logo}
seqLogo(pwm)
```

### Match matrix

Now scan the genome for the motif and see how many are in TEs.

```{r scan}
matches <- matchPWM(pwm, Athaliana, "85%")
length(matches)
```

So we have `r length(matches)` instances of our motif in the Arabidopsis genome, using a cutoff of 85% minimum score.

Now let's check how many are within TEs. First we need to construct our TE genomic ranges object:

### Intersect TEs

```{r read_tes}
te <- import("../../TEPID/Annotation/Arabidopsis/TAIR9_TE.bed.gz",
             format = "bed",
             extraCols = c(strand = "character",
                           TE = "character",
                           family = "character",
                           superfamily = "character"))
# make chromosome names match the genome
te <- renameSeqlevels(te, c("Chr1", "Chr2", "Chr3", "Chr4", "Chr5"))
```


```{r}
n_motif_in_te <- length(subsetByOverlaps(matches, te, ignore.strand = TRUE)) / length(matches)
te_intersections <- subsetByOverlaps(te, matches, ignore.strand = TRUE)
```

Interestingly, we see that `r round(n_motif_in_te, 2) * 100`% of all E2F DEL1 sites are within TEs.

We should see that certain families are overrepresented (SIMPLEHAT, SIMPLEGUY etc.)

```{r}
n_motifs_te <- length(te_intersections)
family_enrichment <- as_data_frame(te_intersections) %>%
  group_by(family) %>%
  mutate(count = n(), proportion = count / n_motifs_te * 100) %>%
  select(count, proportion, family) %>%
  unique()

n_te <- length(te)
all_te <- as_data_frame(te) %>%
  group_by(family) %>%
  mutate(total_count = n(), prop_family = total_count / n_te * 100) %>%
  select(total_count, prop_family, family, superfamily) %>%
  unique()

# join
joined_data <- left_join(all_te, family_enrichment)
joined_data[is.na(joined_data)] <- 0
joined_data <- mutate(joined_data, enrichment = log2((proportion + 1) / (prop_family + 1)))

# just look at the top families
subset_top_families <- arrange(joined_data, desc(enrichment)) %>% head(15)
ggplot(subset_top_families, aes(family, enrichment, fill = superfamily)) + geom_bar(stat = "identity") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

Matches up with the families reported in the Henaff paper.

Now look at the superfamilies.

```{r}
sfamily_enrichment <- as_data_frame(te_intersections) %>%
  group_by(superfamily) %>%
  mutate(count = n(), proportion = count / n_motifs_te * 100) %>%
  select(count, proportion, superfamily) %>%
  unique()

n_te <- length(te)
all_te <- as_data_frame(te) %>%
  group_by(superfamily) %>%
  mutate(total_count = n(), prop_sfamily = total_count / n_te * 100) %>%
  select(total_count, prop_sfamily, superfamily) %>%
  unique()

# join
sfam_joined_data <- left_join(all_te, sfamily_enrichment)
sfam_joined_data[is.na(sfam_joined_data)] <- 0
sfam_joined_data <- mutate(sfam_joined_data, enrichment = log2((proportion + 1) / (prop_sfamily + 1)))

ggplot(sfam_joined_data, aes(superfamily, enrichment)) + geom_bar(stat = "identity") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```


Importantly, we get different results when using peaks vs when scanning the genome for the motif. It seems like using the motif will give better results when looking at TE sequences, as it avoids the problem of multimapped reads. It will also allow analysis of other genomes, not just Arabidopsis, assuming the motif is conserved between the species (could look at *Brassica*). For now, I will proceed with the motif data. The process will be mostly the same regardless of whether a motif or peak is used, as it will only change how the TEs are identified, and so it shouldn't be too difficult to change / compare later.

We can also check more formally for an enrichment of motifs in the TE sequences by calculating the odds ratio.

## Odds ratio

First construct a small matrix that will hold information about the number of bases in / not in the motifs. We then use Fisher's exact test to find the odds ratio.

```{r}
# construct matrix
inOut <- matrix(0, ncol = 2, nrow = 2)
colnames(inOut) <- c("in", "out")
rownames(inOut) <- c("in", "out")
inOut
```

Fill with values

```{r}
inOut[1,1] <- sum(width(GenomicRanges::intersect(matches, te, ignore.strand = TRUE)))
inOut[1,2] <- sum(width(GenomicRanges::setdiff(matches, te, ignore.strand = TRUE)))
inOut[2,1] <- sum(width(GenomicRanges::setdiff(te, matches, ignore.strand = TRUE)))
inOut
```

```{r}
inOut[2,2] <- sum(seqlengths(Athaliana)) - sum(inOut)
inOut
```

odds ratio

```{r fisher_test}
f <- fisher.test(inOut)
f$estimate
```

Log odds ratio

```{r}
log10(f$estimate)
```

Clearly there is a huge enrichment within TEs.
